### PR TITLE
lua: merge lua5.X interpreters

### PR DESCRIPTION
**Pull request by teto**
similar to what was done for python.
Makes it easier to change the passthru settings and the lua infrastructure.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by teto** [comment]
the expansion of makeFlagsArray is awkward. Seems to work locally :/

**Comment by teto** [comment]
@7c6f434c what do you think ? it's not more elegant than the previous implementation but it makes it easier to keep lua interpreters in sync (especially the passthru part with all the buildEnv logic).

File: pkgs/development/interpreters/lua-5/interpreter.nix [hunk 1]
```diff
@@ -22,29 +27,39 @@ stdenv.mkDerivation rec {
...
-    makeFlagsArray=( INSTALL_TOP=$out INSTALL_MAN=$out/share/man/man1 PLAT=linux CFLAGS="-DLUA_USE_LINUX -O2 -fPIC${if compat then " -DLUA_COMPAT_ALL" else ""}" LDFLAGS="-fPIC" V=${luaversion} R=${version} CC="$CC" AR="$AR q" RANLIB="$RANLIB" )
-    installFlagsArray=( TO_BIN="lua luac" TO_LIB="liblua.a liblua.so liblua.so.${luaversion} liblua.so.${version}" INSTALL_DATA='cp -d' )
-    cat ${./lua-5.3-dso.make} >> src/Makefile
-    sed -e 's/ALL_T *= */& $(LUA_SO)/' -i src/Makefile
-  '';
+  makeFlags = [
+    "INSTALL_TOP=${placeholder "out"}"
+    "INSTALL_MAN=${placeholder "out"}/share/man/man1"
+    "R=${version}"
+    ''LDFLAGS="-fPIC"''
```
Comment (7c6f434c) [inline pkgs/development/interpreters/lua-5/interpreter.nix]: Maybe add  comment why these quotes are needed (and why CFLAGS cannot be handled like that)?

**Comment by 7c6f434c** [comment]
Seems fine to me. `a grep a` is a FIXME marker? Does that part work now?

File: pkgs/development/interpreters/lua-5/default.nix [hunk 2]
```diff
@@ -29,44 +25,36 @@ in rec {
       sed -e 's/ALL_T *= */& $(LUA_SO)/' -i src/Makefile
     '';
 
-    # a grep a
     postBuild = stdenv.lib.optionalString (!stdenv.isDarwin) ''
       set -x
```
Comment (7c6f434c) [inline pkgs/development/interpreters/lua-5/default.nix]: Is there a reason to keep verbosity here?

Comment (teto) [reply pkgs/development/interpreters/lua-5/default.nix]: absolutely none sorry :/

Comment (7c6f434c) [reply pkgs/development/interpreters/lua-5/default.nix]: Mwahaha that's what happens when you remove FIXME without looking around why it was there! (Not that it matters)

Comment (teto) [reply pkgs/development/interpreters/lua-5/default.nix]: Yep, I usually leave a TODO but the `a grep a` looked like a mispaste. It also happens when I work on a PR over longer than expected.
I believe it's ready to squash & merge. Not sure if it should go to staging or master though.

**Comment by teto** [comment]
@GrahamcOfBorg build lua5_2 lua5_2_compat lua5_3 lua5_3_compat lua5_1 luaPackages.lpeg

File: pkgs/development/interpreters/lua-5/default.nix (hunk 2 above)
Comment (teto) [reply pkgs/development/interpreters/lua-5/default.nix]: arf it wasn't. Damn shell expansion rules ! reverted my cosmetic-induced bug.

**Comment by 7c6f434c** [comment]
@GrahamcOfBorg build lua5_2 lua5_2_compat lua5_3 lua5_3_compat lua5_1 luaPackages.lpeg

**Comment by 7c6f434c** [comment]
Wait, so quoting works for CC? Would it also work for CFLAGS then?

**Comment by teto** [comment]
In current nixpkgs, lua5.1 has an extra `hydraPlatforms = stdenv.lib.platforms.linux;` compared to this PR. Should I keep it ? what is that and why only 5.1 has it; not 5.2 or 5.3 ?

**Comment by 7c6f434c** [comment]
I would guess that in old days Darwin was even lower tier of support, so somebody wanted to remove the platforms restriction without needing to consider Hydra impact. Feel free to git blame to find out, of course.

**Comment by teto** [comment]
@GrahamcOfBorg build lua5_2 lua5_2_compat lua5_3 lua5_3_compat lua5_1 luaPackages.lpeg

**Comment by teto** [comment]
@GrahamcOfBorg build lua5_2 lua5_2_compat lua5_3 lua5_3_compat lua5_1 luaPackages.lpeg

**Comment by flokli** [comment]
This seems to have broken the ultrastardx build, at least according to a git bisect:

```
Linking ../game/ultrastardx
Warning: "crti.o" not found, this will probably cause a linking failure
Warning: "crtn.o" not found, this will probably cause a linking failure
/nix/store/2dfjlvp38xzkyylwpavnh61azi0d168b-binutils-2.31.1/bin/ld: warning: ../game/link.res contains output sections; did you forget -T?
/nix/store/2dfjlvp38xzkyylwpavnh61azi0d168b-binutils-2.31.1/bin/ld: cannot find -llua5.2
Error: Error while linking
Fatal: There were 1 errors compiling module, stopping
Fatal: Compilation aborted
Error: /nix/store/6y49zi0s1s7fwirv4yd3835khnmh0wl0-fpc-3.0.0/bin/ppcx64 returned an error exitcode
make[1]: *** [Makefile:233: ../game/ultrastardx] Error 1
make[1]: Leaving directory '/build/source/src'
make: *** [Makefile:136: all] Error 2
builder for '/nix/store/kp3i70044s81b746c733201wsjl4x8s8-ultrastardx-2017.8.0.drv' failed with exit code 2
```
(nix log /nix/store/kp3i70044s81b746c733201wsjl4x8s8-ultrastardx-2017.8.0.drv)
https://hydra.nixos.org/build/93748657/nixlog/1

**Comment by teto** [comment]
I can reproduce, I'll look into it.

**Comment by teto** [comment]
I think the flags are wrong, looking at lua5.2 pkgconfig, it advises to link with -llua, not -llua5.2:
`Libs: -L/nix/store/l00fyl6fqhgbw2rg3nkw0qwd3jh32ibp-lua-5.2.4/lib -llua -lm
Maybe the maintainer could bump the package to today's version as there seems to be activity but no release on the github https://github.com/UltraStar-Deluxe/USDX
`

**Comment by flokli** [comment]
@teto I'm not sure if I understand the problem. lua 5.2 pkgconfig did say `-llua` before and after 672c3c1, so this can't be the reason for the build to fail.

The `ultrastardx` package is already at the latest release, and there seem to be no lua-related changes in the diff.

**Comment by teto** [comment]
Shouldn't they use `pkg-config --libs lua` to get the flags instead ?

**Comment by flokli** [comment]
they seem to do: https://github.com/UltraStar-Deluxe/USDX/blob/114ff0c11e8f2c415404e75db213f97a744f487b/configure.ac#L196-L218

**Comment by teto** [comment]
I've opened an issue here https://github.com/UltraStar-Deluxe/USDX/issues/462

**Comment by flokli** [comment]
@teto I'm not sure I understand how this PR could have broken it, and why it did work before.

**Pull request merged**